riscv-page-tables/src/page_table.rs

```diff
@@ -1098,3 +1064,37 @@ impl<'a, T: PagingMode> Drop for GuestStageMapper<'a, T> {
         }
     }
 }
+
+pub struct InvalidateIter<'a, T: PagingMode> {
+    inner: MutexGuard<'a, PageTableInner<T>>,
+    vaddr: PageAddr<T::MappedAddressSpace>,
+    end_vaddr: PageAddr<T::MappedAddressSpace>,
+}
+
+impl<'a, T: PagingMode> InvalidateIter<'a, T> {
+    fn new(
+        inner: MutexGuard<'a, PageTableInner<T>>,
+        vaddr: PageAddr<T::MappedAddressSpace>,
+        end_vaddr: PageAddr<T::MappedAddressSpace>,
+    ) -> Self {
+        Self {
+            inner,
+            vaddr,
+            end_vaddr,
+        }
+    }
+}
+
+impl<'a, T: PagingMode> Iterator for InvalidateIter<'a, T> {
+    type Item = SupervisorPageAddr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.vaddr == self.end_vaddr {
+            return None;
+        }
+        let pte = self.inner.get_mapped_4k_leaf(self.vaddr).ok()?.invalidate();
```

Comment:
This can't happen; we already checked the entire range can be invalidated and we've held an exclusive lock the whole time. Changed it to `unwrap()` with a comment.